README.md
domain_name_validator
=====================

Ever needed to validate a domain name? This gem will validate any domain name
represented in ASCII.

The scope of this gem is deliberately focused on validating domain names. It
simply answers the question: "Is this a real domain name?" Using this command,
you can make a realistic assessment about whether you want to store a domain
name or URL in your database. This gem will tell you 1) that a domain is or
is not valid, and 2) if it's not valid, what the errors are. 

There are other gems and libraries that parse domain names into their various
components, or parse URLS, or properly handle Unicode domain names, etc. Use
them; many of them are very good at their well-defined roles. But none of the
ones that I came across were very good at simply telling me whether a domain
names was valid, or WHY it was invalid if it failed the validations.

How It Works
------------

To validate a domain name:

    v = DomainNameValidator.new
    if v.validate('keenertech.com')
      # Do something
    end

What about error messages? If a domain isn't valid, it's often desirable to
find out why the domain ewasn't valid. To do this, simply pass an array into
the "validate" message as the optional second argument.

    errs = []
    v = DomainNameValidator.new
    unless v.validate('keenertech.123', errs)
      puts("Errors: #{errs.inspect}")
    end

This generates the following output:

    Errors: ["The top-level domain (the extension) cannot be numerical"]

This gem should make it easy to validate domain names.

About Domain Names
------------------

Domain names provide a unique, memorizable name to represent numerically
addressable Internet resources. They also provide a level of abstraction that
allows the underlying Internet address to be changed while still referencing
a resource by its domain name. The domain name space is managed by the 
Internet Corporation for Assigned Names and Numbers (ICANN).

The right-most label of a domain name is referred to as the top-level domain,
or TLD. A limited set of top-level domain names, and two-character country
codes, have been standardized. The Internet Assigned Numbers Authority (IANA)
maintains an annotated list of top-level domains, as well as a list of
"special use," or reserved, top-level domain names.

* http://www.iana.org/domains/root/db/
* http://www.iana.org/assignments/special-use-domain-names/special-use-domain-names.xml

Domain names follow some very detailed rules:

* The maximum length of a domain name is 253 characters.

* A domain name is divided into "labels" separated by periods. The maximum
  number of labels is 127.

* The maximum length of any label within a domain name is 63 characters.

* No label, including TLDs, can begin or end with a dash.

* Top-level domain names cannot be all numeric.

* The right-most label must be either a recognized TLD or a 2-letter country
  code. The only exception is for international domain names, for which
  TLD checking is currently bypassed.

* Top-level domain names cannot be all numeric.

* Domain names may not begin with a period.


Internationalized Domain Names
------------------------------

What about internationalized domain names? ICANN approved the Internationalized
Domain Name (IDNA) system in 2003. This standard allows for Unicode domain
names to be encoded into ASCII using Punycode. Essentially, a label may contain
"xn--" as a prefix, followed by the Punycode representation of a Unicode string,
resulting in domain names such as xn--kbenhavn-54.eu. Note that there are also
some approved Unicode TLDs.

The process of rendering an internationalized domain name in ASCII via 
Punycode is called normalization. This gem will validate a normalized domain
name, but not a Unicode domain name. Note, however, that it currently does not
validate normalized TLDs against ICANN's list of valid TLDs.

It's also unclear whether the "xn--" prefix should count against the label
size limit of 63 characters. In the absence of specific guidelines, and because
I've never actually seen an overly long label, I have chosen to apply the limit
irregardless of the presence of the "xn--" prefix within a label.

Requirements
------------

This is a Ruby gem with no run-time dependencies on anything else. It's only
been tested under Ruby 1.9.3, but it should be compatible with all versions of
Ruby more recent than Ruby 1.8.6.

Install
-------

Installation doesn't get much simpler than this:

    gem install domain_name_validator

Road Map
--------

More types of checks will be added as they are identified. Support for
validating top-level domains (TLD's) is also in the works (it's a bit more
complex than you might imagine). This will also bring in the capability to
validate effective TLD's as well, i.e. - strings that are treated as top-level
domains by registrars around the world.

Alternative Gems
----------------

If this domain_name_validator gem does not suit your needs, here are a few 
recommended gems that may provide you with the additional power (and
complexity) that is deliberately absent from this highly focused gem:

* domain_name - A full-featured gem for parsing/manipulating domain names.
* ip_address - For everything you need to do with Ipv4 and Ipv6 addresses.


Author
------

David Keener

He's a long-time Rubyist, with extensive experience both in the Internet
startup world and government contracting. He is one of the founders of the
RubyNation, DevIgnition and NationJS conferences. He speaks often at technical
conferences, and blogs regularly on Internet-related subjects at
KeenerTech.com.

Contributors
------------

Many thanks for the support of General Dynamics and the Department of
Homeland Security (DHS). And more specifically on input from Andrew Finch,
Josh Lentz, Jonathan Quigg and Dave Roberts.

YOUR SUPPORT
------------

Please help me make this gem as useful as possible for its admittedly limited
purpose. If ICANN's rules change, a clearer rules interpretation becomes
available, or new valid TLDs are added, please help out by notifying me of
potentially useful changes, or by submitting a pull request via GitHub.


ChangeLog.md
0.5 (2013-10-24)
----------------

* By request, the gem now ignores leading/trailing whitespace when validating
  domain names. This includes ignoring newlines at the end of domain names.

* Remove timestamp from gemspec so Rubygems will calculate gem dates for us.

* (0.5.1 - 2013-11-06) The validate method now avoids accidental memoization
  on the part of the caller by ensuring that the contents of the array
  argument have been cleared. Previously, if the caller called the method
  multiple times using the same array, then all calls would result in FALSE
  after the first FALSE was legitemately encountered. Yes, somebody did
  this.

0.4 (2013-07-17)
----------------

* By request, added rudimentary TLD checking. A TLD, the right-most label of a
  domain name, should be either 2 or 3 characters, unless it's "aero", "arpa",
  "museum" or begins with "xn--" (for a normalized Unicode TLD). This check
  does not validate against a master list of TLD's, but basically just fails
  a domain name if the TLD basically could not be valid under any
  circumstances. This validation would, for example, reject a domain name
  such as "test.domain".

* (0.4.1 - 2013-07-17) Added "info" as a valid extra-long TLD. How I missed
  this one, I have no idea.

* (0.4.2 - 2013-07-17) Added a check for zero-length domain names, either
  empty strings or Ruby nil values being passed in. In practice, these edge
  cases were actually happening.

* (0.4.3 - 2013-08-01) For the TLD reality check, IANA (the Internet Assigned
  Numbers Authority) recognizes some long TLD's that were not on the original
  list for this gem, such as ".travel", ".jobs", etc. To reiterate, the gem
  does not validate against the full list of TLD's or effective TLD's, but 
  it does validate that the TLD length COULD be a valid TLD. As an example,
  a TLD of 14 characters would be rejected because there are no valid domains
  that could match that pattern.

* (0.4.3 - 2013-08-01) Added the "license" attribute to the gemspec. This is
  in response to an issue that was opened on GitHub.

* (0.4.4 - 2013-08-28) Fixed bug involving proper handling of domain names with
  capital letters. Reported by a user.

0.3 (2013-06-28)
----------------

* Added a check, because labels cannot begin with a period.
* Updated documentation, plus defined Road Map for future changes.

0.2 (2013-06-17)
----------------

First version built and deployed as a gem on a real project. Released on
Rubygems.org for the first time.

* Added more RSpec tests for improved test coverage.
* Enhanced documentation.


0.1 (2013-06-10)
----------------

Initial working version of the code, with minimal RSpec tests. Released on
GitHub for review, but not yet published as a gem to the Ruby community.


spec/spec_helper.rb
require File.expand_path(
    File.join(File.dirname(__FILE__), %w[.. lib domain_name_validator]))
        
RSpec.configure do |config|
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
end


spec/domain_name_validator_spec.rb
require 'spec_helper'

describe DomainNameValidator do

  before(:each) do
    @validator = DomainNameValidator.new
  end

  describe 'Basic Validation' do

    it 'should pass a valid domain name' do
      response = @validator.validate('keenertech.com')
      response.should be == true
    end

    it 'should pass a valid domain name with uppercase letters' do
      response = @validator.validate('KeenerTech.COM')
      response.should be == true
    end

    it 'should fail when it finds nil instead of a domain name' do
      response = @validator.validate(nil)
      response.should be == false
    end

    it 'should fail when it finds an empty string instead of a domain name' do
      response = @validator.validate("")
      response.should be == false
    end

    it 'should pass when it finds a domain with leading/trainling whitespace' do
      response = @validator.validate("  www.domain-1.com\n")
      response.should be == true
    end

    it 'should fail when it finds a numeric top-level extension' do
      response = @validator.validate('keenertech.123')
      response.should be == false
    end

    it 'should fail when the domain name max size is exceeded' do
      domain = "a"*250 + ".com"    # 254 chars; max is 253
      response = @validator.validate(domain)
      response.should be == false
    end

    it 'should fail when the max number of levels is exceeded' do
      domain = "a."*127 + "com"    # 128 levels; max is 127
      response = @validator.validate(domain)
      response.should be == false
    end

    it 'should fail when a label exceeds the max label length' do
      domain = "a"*64 + "b.c.com"    # 64 chars; max is 63 for a label
      response = @validator.validate(domain)
      response.should be == false
    end

    it 'should fail when a label begins with a dash' do
      domain = "a.-b.c.com"
      response = @validator.validate(domain)
      response.should be == false
    end

    it 'should fail when a TLD begins with a dash' do
      domain = "a.b.c.-com"
      response = @validator.validate(domain)
      response.should be == false
    end

    it 'should fail when a domain name begins with a period' do
      domain = ".b.c.com"
      response = @validator.validate(domain)
      response.should be == false
    end
  end

  describe 'Internationalized (normalized) domain names' do

    it 'should pass with a normalized international domain name' do
      domain = "xn--kbenhavn-54.eu"
      response = @validator.validate(domain)
      response.should be == true
    end

  end

  describe 'Rudimentary TLD Checking' do

    it 'should fail if the TLD is too short' do
      domain = "test.a"
      response = @validator.validate(domain)
      response.should be == false
    end

    it 'should fail if the TLD is too long' do
      domain = "test.domain"
      response = @validator.validate(domain)
      response.should be == false
    end

    it "should succeed if the TLD is too long but equals 'aero'" do
      domain = "test.aero"
      response = @validator.validate(domain)
      response.should be == true
    end

    it "should succeed if the TLD is too long but equals 'info'" do
      domain = "test.info"
      response = @validator.validate(domain)
      response.should be == true
    end

    it "should succeed if the TLD is too long but equals 'arpa'" do
      domain = "test.arpa"
      response = @validator.validate(domain)
      response.should be == true
    end

    it "should succeed if the TLD is too long but equals 'museum'" do
      domain = "test.museum"
      response = @validator.validate(domain)
      response.should be == true
    end

    it "should succeed if the TLD is too long but matches 'xn--'" do
      domain = "test.xn--really-long-text"
      response = @validator.validate(domain)
      response.should be == true
    end

  end

end


lib/domain_name_validator/validator.rb
# The purpose of this class is to provide a simple capability for validating
# domain names represented in ASCII, a feature that seems to be missing or
# obscured in other more wide-ranging domain-related gems.

class DomainNameValidator

  MAX_DOMAIN_LENGTH = 253
  MAX_LABEL_LENGTH = 63
  MAX_LEVELS = 127
  MAX_TLD_LENGTH = 3       # Except for "aero", "arpa", "info" and "museum"
  MIN_LEVELS = 2
  MIN_TLD_LENGTH = 2

  ERRS = {
    :bogus_tld =>
       'Malformed TLD: Could not possibly match any valid TLD',
    :illegal_chars =>
       'Domain label contains an illegal character',
    :illegal_start =>
       'No domain name may start with a period',
    :label_dash_begin =>
       'No domain label may begin with a dash',
    :label_dash_end =>
       'No domain label may end with a dash',
    :max_domain_size =>
       'Maximum domain length of 253 exceeded',
    :max_label_size =>
       'Maximum domain label length of 63 exceeded',
    :max_level_size =>
       'Maximum domain level limit of 127 exceeded',
    :min_level_size =>
       'Minimum domain level limit of 2 not achieved',
    :top_numerical =>
       'The top-level domain (TLD) cannot be numerical',
    :zero_size =>
       'Zero-length domain name'
    }

  # Validates the proper formatting of a normalized domain name, i.e. - a
  # domain that is represented in ASCII. Thus, international domain names are
  # supported and validated, if they have undergone the required IDN
  # conversion to ASCII. The validation rules are:
  #
  # 1. The maximum length of a domain name is 253 characters.
  # 2. A domain name is divided into "labels" separated by periods. The maximum
  #    number of labels (including the top-level domain as a label) is 127.
  # 3. The maximum length of any label within a domain name is 63 characters.
  # 4. No label, including top-level domains, can begin or end with a dash.
  # 5. Top-level names cannot be all numeric.
  # 6. A domain name cannot begin with a period.

  def validate(dn, errs = [])
    errs.clear   # Make sure the array starts out empty
    if dn.nil?
      errs << ERRS[:zero_size]
    else
      dn = dn.strip
      errs << ERRS[:zero_size] if dn.size == 0
    end

    if errs.size == 0
      errs << ERRS[:max_domain_size] if dn.size > MAX_DOMAIN_LENGTH
      parts = dn.downcase.split('.')
      errs << ERRS[:max_level_size] if parts.size > MAX_LEVELS
      errs << ERRS[:min_level_size] if parts.size < MIN_LEVELS
      parts.each do |p|
        errs << ERRS[:max_label_size] if p.size > MAX_LABEL_LENGTH
        errs << ERRS[:label_dash_begin] if p[0] == '-'
        errs << ERRS[:label_dash_end] if p[-1] == '-'
        errs << ERRS[:illegal_chars] unless p.match(/^[a-z0-9\-\_]+$/)
      end
      errs << ERRS[:top_numerical] if parts.last.match(/^[0-9]+$/)
      if parts.last.size < MIN_TLD_LENGTH || parts.last.size > MAX_TLD_LENGTH
        unless parts.last == 'arpa' ||
               parts.last == 'aero' ||
               parts.last == 'asia' ||
               parts.last == 'coop' ||
               parts.last == 'info' ||
               parts.last == 'jobs' ||
               parts.last == 'mobi' ||
               parts.last == 'museum' ||
               parts.last == 'name' ||
               parts.last == 'post' ||
               parts.last == 'travel' ||
               parts.last.match(/^xn--/)
          errs << ERRS[:bogus_tld]   
        end
      end
      errs << ERRS[:illegal_start] if parts.first[0] == '.'
    end

    errs.size == 0   # TRUE if valid, FALSE otherwise
  end

end
